src/main/java/net/snowflake/client/core/SFArrowResultSet.java

```diff
@@ -443,6 +448,7 @@ public Date getDate(int columnIndex, TimeZone tz) throws SFException {
     int index = currentChunkIterator.getCurrentRowInRecordBatch();
     wasNull = converter.isNull(index);
     converter.setSessionTimeZone(timeZone);
+    converter.setUseSessionTimezone(useSessionTimezone);
```

Comment:
Do we need to have converters to do the time zone offset? We have actual time/date/ts object here and we have the flag. Could we directly wrap the returned value to the type we want?

Reply:
Please take a look at the Arrow code folder here: [https://github.com/snowflakedb/snowflake-jdbc/tree/master/src/main/java/net/snowflake/client/core/arrow](https://github.com/snowflakedb/snowflake-jdbc/tree/master/src/main/java/net/snowflake/client/core/arrow). The way it's set up is that for each fetching function (getDate(), getTimestamp(), etc), the converter of type ArrowVectorConverter is initialized as a child class of this Converter class depending on the SQL data type. Each converter child class has a different implementation of each method fetching method. So returning the time/date/ts object directly here would break this implementation. The objects are returned in the implementations of the functions in the child classes, and the converter can have relevant values set with these setSessionTimeZone/ isNull functions.

Reply:
What I meant is something like ToSessionTime(converter.toDate(index, tz, resultSetSerializable.getFormatDateWithTimeZone()), take the results returned from converter THEN process with session offset.

Reply:
And here we already know the type is Date, for unknown type case, e.g. getObject case, we could detect if it is the time types of interests.

Reply:
Ah, I see. I prefer using the converter.set functions rather than adding parameters to functions like getDate, getTime, etc. The reason is that we will likely need to add more parameters in the future, and this will get confusing because 1.) there will be so many function parameters that it's difficult to keep track when trying to use the function, and 2.) with each new addition of a parameter, the entire function signature and every usage of that function in the code would have to be refactored, which is messy. 
Also, not all of the parameters are required for each call of the function, which would lead to a lot of null inputs if each parameter was included in the function signature. For example, the session timezone is not required for when getString() calls the toDate() method. With the current implementation, the converter set function is simply not called. If it was changed so that toDate() now has type toDate(boolean useSessionTimezone), then the getString() method will likely call toDate(null) because there's no timezone, and it's good practice to avoid using null when possible to avoid accidental nullptrexceptions.

Reply:
I think we don't need to add parameters to getDate/getTime. The 'session' and 'resultSetSerializable' objects are in SFBaseResultSet already (BTW, I don't think we need useSessionTimeZone in resultSetSerializable. Since session object is always available when resultSetSerializable.useSessionTimeZone is used). We could just refer to any parameter needed from these two objects. 

When we think about this session time zone task in a higher view, Ideally we should not need to touch the code that is not handling time/date/ts types, and since this functionality is not JSON or Arrow specific, ideally the parameter checking and actual timezone offset code should not happen in the code specific to them or converters (very Arrow specific). So I think ArrowResultUtil (methods here are not Arrow specific anyway) and ResultUtil could be the places to do the actual conversion(getSFTimeAsString, etc, we need to pass session and/or resultSetSerializable to them) and we should be able to solve the getString() case you mentioned and also avoid null calls. 

Overall I hope this could be a simpler solution and less intrusive to the Arrow or Json specific code.

Reply:
This is true, I think the entire ResultSet system could use an overhaul. There are several session parameters (JDBC_TREAT_NTZ_AS_UTC, JDBC_FORMAT_DATE_WITH_TIMEZONE, CLIENT_HONOR_CLIENT_TZ_FOR_TIMESTAMP_NTZ, etc) that should also be refactored at the same time because the logic between these parameters is somewhat intertwined so all these changes should be made together so as to not break anything. I am filing a ticket to look into a better design: [https://snowflakecomputing.atlassian.net/browse/SNOW-241484](https://snowflakecomputing.atlassian.net/browse/SNOW-241484)

One other note is that the session object is not always available to retrieve the information from useSessionTimezone. When ResultSetSerializable is used, often the session object is null. This is why the value is copied from the session into the ResultSetSerializable instance when it's created.